lib/common/socket.c

```diff
@@ -1589,15 +1615,90 @@ static void on_handshake_fail_complete(h2o_socket_t *sock, const char *err)
 
 static void proceed_handshake(h2o_socket_t *sock, const char *err);
 
+#if H2O_CAN_ASYNC_SSL
+
+static void on_async_proceed_handshake(h2o_socket_t *async_sock, const char *err)
+{
+    if (err != NULL)
+        h2o_fatal("error on internal notification fd:%s", err);
+
+    /* Do we need to handle spurious events for eventfds / pipes used for intra-process communication? If so, we have to read
+     * something here (or let `async_cb` read and return if it succeeded). */
+
+    h2o_socket_t *sock = async_sock->data;
+    assert(sock->ssl->async.inflight);
+    sock->ssl->async.inflight = 0;
+
+    h2o_socket_read_stop(async_sock);
+    dispose_socket(async_sock, NULL);
+
+    proceed_handshake(sock, NULL);
+}
+
+static void do_proceed_handshake_async(h2o_socket_t *sock, ptls_buffer_t *ptls_wbuf)
+{
+    assert(!sock->ssl->async.inflight);
+    sock->ssl->async.inflight = 1;
+    h2o_socket_read_stop(sock);
+
+    /* get async fd and retain wbuf */
+    int async_fd;
+    if (sock->ssl->ptls != NULL) {
+        async_fd = ptls_openssl_get_async_fd(sock->ssl->ptls);
+        sock->ssl->async.ptls_wbuf = *ptls_wbuf;
+        *ptls_wbuf = (ptls_buffer_t){NULL};
+    } else {
+        assert(ptls_wbuf == NULL);
+        size_t numfds;
+        SSL_get_all_async_fds(sock->ssl->ossl, NULL, &numfds);
+        assert(numfds == 1);
+        SSL_get_all_async_fds(sock->ssl->ossl, &async_fd, &numfds);
+        assert(numfds == 1);
+    }
+
+    /* dup async_fd as h2o socket handling will close it */
+    if ((async_fd = dup(async_fd)) == -1) {
+        char errbuf[256];
+        h2o_fatal("dup failed:%s", h2o_strerror_r(errno, errbuf, sizeof(errbuf)));
+    }
+
+    /* add async fd to event loop in order to retry when openssl engine is ready */
+#if H2O_USE_LIBUV
+    h2o_socket_t *async_sock = h2o_uv__poll_create(h2o_socket_get_loop(sock), async_fd, (uv_close_cb)free);
+#else
+    h2o_socket_t *async_sock = h2o_evloop_socket_create(h2o_socket_get_loop(sock), async_fd, H2O_SOCKET_FLAG_DONT_READ);
+#endif
+    async_sock->data = sock;
+    h2o_socket_read_start(async_sock, on_async_proceed_handshake);
+}
+
+#endif
+
 static void proceed_handshake_picotls(h2o_socket_t *sock)
 {
     size_t consumed = sock->ssl->input.encrypted->size;
     ptls_buffer_t wbuf;
-    ptls_buffer_init(&wbuf, "", 0);
+
+#if H2O_CAN_ASYNC_SSL
+    if (sock->ssl->async.ptls_wbuf.base != NULL) {
+        wbuf = sock->ssl->async.ptls_wbuf;
+        sock->ssl->async.ptls_wbuf = (ptls_buffer_t){NULL};
+    } else
+#endif
+        ptls_buffer_init(&wbuf, "", 0);
```

Comment:
nit: indent, clang-format might catch this

Reply:
clang-format on my MacBook was fine with this, but I agree that the use of #if is quirky here. Fixed in 4c12be4.